azurelinuxagent/common/protocol/hostplugin.py

```diff
@@ -117,6 +147,44 @@ def get_artifact_request(self, artifact_url, artifact_manifest_url=None):
 
         return url, headers
 
+    def report_fetch_health(self, uri, is_healthy=True, source='', response=''):
+
+        if uri != URI_FORMAT_GET_EXTENSION_ARTIFACT.format(self.endpoint, HOST_PLUGIN_PORT):
+            return
+
+        if self.should_report(is_healthy,
+                              self.fetch_error_state,
+                              self.fetch_last_timestamp,
+                              HostPluginProtocol.FETCH_REPORTING_PERIOD):
+            self.fetch_last_timestamp = datetime.datetime.utcnow()
+            health_signal = self.fetch_error_state.is_triggered() is False
+            self.health_service.report_host_plugin_extension_artifact(is_healthy=health_signal,
+                                                                      source=source,
+                                                                      response=response)
+
+    def report_status_health(self, is_healthy, response=''):
+        if self.should_report(is_healthy,
+                              self.status_error_state,
+                              self.status_last_timestamp,
+                              HostPluginProtocol.STATUS_REPORTING_PERIOD):
+            self.status_last_timestamp = datetime.datetime.utcnow()
+            health_signal = self.status_error_state.is_triggered() is False
+            self.health_service.report_host_plugin_status(is_healthy=health_signal,
+                                                          response=response)
+
+    @staticmethod
+    def should_report(is_healthy, error_state, last_timestamp, period):
+
+        if is_healthy:
+            error_state.reset()
+        else:
+            error_state.incr()
+
+        if last_timestamp is None:
+            last_timestamp = datetime.datetime.utcnow() - period
+
+        return datetime.datetime.utcnow() >= (last_timestamp + period)
```

Comment:
>  return datetime.datetime.utcnow() >= (last_timestamp + period) [](start = 6, length = 64)

If you report, should you reset and start over again for the next period?  Is that handled elsewhere?

Reply:
No, we need to keep reporting failure until we see success. This is handled by line 179. This is a little different to how the existing error states are used.

Reply:
Sure.